libs/core/algorithms/include/hpx/algorithms/traits/is_trivially_relocatable.hpp

```diff
@@ -0,0 +1,38 @@
+//  Copyright (c) 2023 Isidoros Tsaousis-Seiras
+//
+//  SPDX-License-Identifier: BSL-1.0
+//  Distributed under the Boost Software License, Version 1.0. (See accompanying
+//  file LICENSE_1_0.txt or copy at http://www.boost.org/LICENSE_1_0.txt)
+
+#pragma once
+
+#include <hpx/iterator_support/traits/is_iterator.hpp>
+#include <type_traits>
+
+// Macro to specialize template for given type
+#define HPX_DECLARE_TRIVIALLY_RELOCATABLE(T)                                   \
+    namespace hpx {                                                            \
+        template <>                                                            \
+        struct is_trivially_relocatable<T> : std::true_type                    \
+        {                                                                      \
```

Comment:
last minute thought: adding static_assert(is_relocatable<T>) ; as it is impossible to be trivially_relocatable but not relocatable.

Reply:
That's what I thought as of P1144R6; but the current revision (P1144R8) adopts the viewpoint that "trivially relocatable" is like "trivially copyable." It implies, "I promise that **hypothetically when** you (copy|relocate|compare) this thing, you can do it trivially." It doesn't imply anything about whether it is **actually okay** to (copy|relocate|compare) the thing; that's up to the algorithm's author to check.

E.g.
```
struct A { int i; A(A&&) = delete; A& operator=(A&&) = delete; };
static_assert(std::is_trivially_copyable_v<A>);  // C++ today
static_assert(std::is_trivially_relocatable_v<A>);  // in P1144R8
A src[10];
A dst[10];
alignas(A) char rawdst[10*sizeof(A)];
std::copy(src, src+10, dst);  // invalid C++ today
std::uninitialized_move(src, src+10, (A*)rawdst);  // also invalid C++ today
std::uninitialized_relocate(src, src+10, (A*)rawdst);  // also invalid in P1144R8
```

...however, I see that I need to update my libc++ implementation to actually reject the latter. ;)

Anyway, I don't have a strong opinion on which way HPX should define their trait. If you want to make it false for non-relocatable types, that's not crazy; it won't explode or anything. But I did want to point out _that_ P1144R8 chooses to do it differently, and _why_ P1144R8 chooses to do it differently.

Reply:
I see, I am struggling to think of a case where using std::relocate on a trivially relocatable but not relocatable type would be correct. I think I must be missing something

Reply:
> I am struggling to think of a case where using [std::uninitialized_relocate] on a trivially relocatable but not relocatable type would be correct.

That's the point: it'd never be correct. And neither would it be correct to use `std::copy` on a trivially copyable but not copy-assignable type. That doesn't mean such a type is _not_ trivially copyable (or _not_ trivially relocatable); it just means that you can't use it with those algorithms because they require copy-assignability resp. relocatability.

The  _interface constraint_ on `std::copy` is that the type you pass in needs to be `is_copy_assignable`. Internally, the _optimization into a memcpy_ kicks in when the type happens to be `is_trivially_copyable`.

The _interface constraint_ on `std::uninitialized_relocate` is that the type you pass in needs to be `is_relocatable`. Internally, the _optimization into a memcpy_ kicks in when the type happens to be `is_trivially_relocatable`.

Reply:
Ohhh I see, so it's just a matter of where this check should be performed, R8 suggests it should not throw a compile error when you give a non-relocatable type this trait, but it should when we try to relocate it, right? And this is checked with a static assertion inside the relocation algorithms

Reply:
Yes, I believe you've got it now. Here's a concrete example with libc++: https://godbolt.org/z/Yr9MdncoT
```
struct S {
    const std::unique_ptr<int> m;
};

static_assert(std::is_trivially_relocatable_v<S>);
static_assert(!std::is_relocatable_v<S>);

void f(S *p) {
    std::relocate_at(p, p); // ill-formed
}

__memory/relocate_at.h:91:19: error: static assertion failed [...]
```
If you _could_ relocate `S`, then you could do so trivially. But in fact you can't relocate `S`.

Reply:
| If you could relocate `S`, then you could do so trivially. But in fact you can't relocate `S`.

Just for a sanity check; arrays of trivially relocatable types `T[]` should be considered trivially relocatable, but not relocatable (as not move constructable)?

Reply:
> Just for a sanity check; arrays of trivially relocatable types T[] should be considered trivially relocatable, but not relocatable (as not move constructable)?

Yes, that's exactly right. https://godbolt.org/z/9n4YGTjfd  (P1144R8 has actually removed `std::is_relocatable_v`, but kept `std::relocatable`. My Godbolt implementation keeps both, at least for now, since they're trivial to implement.)